### PR TITLE
Feature/transfer-learning

### PR DESCRIPTION
**Pull request by icedoom888**
WIP: Working on implementation of Transfer Learning.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-training?pullRequest=166) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-training?pullRequest=166) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-training?pullRequest=166) it.</sub>

**Comment by icedoom888** [comment]
Addresses: #167

**Comment by icedoom888** [comment]
NOT READY YET

File: src/anemoi/training/config/model/graphtransformer.yaml
Comment (mchantry) [inline src/anemoi/training/config/model/graphtransformer.yaml:1]: Why is this file edited?

File: src/anemoi/training/train/train.py [hunk 1]
```diff
@@ -289,11 +314,12 @@ def _log_information(self) -> None:
         LOGGER.debug("Total number of auxiliary variables: %d", len(self.config.data.forcing))
 
         # Log learning rate multiplier when running single-node, multi-GPU and/or multi-node
-        total_number_of_model_instances = (
+        total_number_of_model_instances = int(
```
Comment (mchantry) [inline src/anemoi/training/train/train.py:317]: This is separate from transfer learning, no?

File: src/anemoi/training/train/train.py [hunk 2]
```diff
@@ -390,7 +420,8 @@ def train(self) -> None:
 
 @hydra.main(version_base=None, config_path="../config", config_name="config")
 def main(config: DictConfig) -> None:
-    AnemoiTrainer(config).train()
+    trainer = AnemoiTrainer(config)
```
Comment (mchantry) [inline src/anemoi/training/train/train.py:423]: Do you need the trainer object?

File: src/anemoi/training/utils/checkpoint.py [hunk 3]
```diff
@@ -65,3 +68,32 @@ def save_inference_checkpoint(model: torch.nn.Module, metadata: dict, save_path:
     torch.save(model, inference_filepath)
     save_metadata(inference_filepath, metadata)
     return inference_filepath
+
+
+def sanify_checkpoint(model: torch.nn.Module, ckpt_path: Path | str, save_path: Path | str) -> Path:
```
Comment (mchantry) [inline src/anemoi/training/utils/checkpoint.py:73]: sanify ? sanitise?

File: src/anemoi/training/train/train.py (hunk 2 above)
Comment (icedoom888) [reply src/anemoi/training/train/train.py:423]: This was a debugging leftover, will remove

File: src/anemoi/training/train/train.py (hunk 1 above)
Comment (icedoom888) [reply src/anemoi/training/train/train.py:317]: Removing this!

File: src/anemoi/training/config/model/graphtransformer.yaml
Comment (icedoom888) [reply src/anemoi/training/config/model/graphtransformer.yaml:1]: This is a mistake, removing these changes, sorry!!

File: src/anemoi/training/train/forecaster.py [hunk 4]
```diff
@@ -529,18 +529,11 @@ def calculate_val_metrics(
                 validation metrics and predictions
         """
         metrics = {}
-
         # Added to impute nans
-        nan_locations = torch.isnan(y[..., self.data_indices.internal_data.output.full])
+        nan_locations = torch.isnan(y)
```
Comment (mchantry) [inline src/anemoi/training/train/forecaster.py:533]: Can you explain this change please?

File: src/anemoi/training/data/dataset.py [hunk 5]
```diff
@@ -55,9 +56,11 @@ def __init__(
             Shuffle batches, by default True
         label : str, optional
             label for the dataset, by default "generic"
-
+        effective_bs : int, default 1
```
Comment (mchantry) [inline src/anemoi/training/data/dataset.py:59]: Why does effective bs need to be introduced for transfer learning?

File: src/anemoi/training/train/forecaster.py [hunk 6]
```diff
@@ -545,9 +548,9 @@ def calculate_val_metrics(
 
             for mkey, indices in self.val_metric_ranges.items():
                 metrics[f"{metric_name}/{mkey}/{rollout_step + 1}"] = metric(
-                    y_pred_postprocessed[..., indices],
-                    y_postprocessed[..., indices],
-                    scalar_indices=[..., indices] if -1 in metric.scalar else None,
+                    y_pred_postprocessed[...,],
```
Comment (mchantry) [inline src/anemoi/training/train/forecaster.py:551]: Why have indices been removed please?

File: src/anemoi/training/train/train.py [hunk 7]
```diff
@@ -34,13 +34,15 @@
...
 from anemoi.training.train.forecaster import GraphForecaster
+from anemoi.training.utils.checkpoint import transfer_learning_loading
 from anemoi.training.utils.jsonify import map_config_to_primitives
 from anemoi.training.utils.seeding import get_base_seed
 
 if TYPE_CHECKING:
     from torch_geometric.data import HeteroData
 
 LOGGER = logging.getLogger(__name__)
+LOGGER.setLevel(logging.DEBUG)  # Change DEBUG to INFO, WARNING, etc., as needed
```
Comment (mchantry) [inline src/anemoi/training/train/train.py:45]: Needs removing when draft phase over?

File: src/anemoi/training/train/forecaster.py (hunk 4 above)
Comment (icedoom888) [reply src/anemoi/training/train/forecaster.py:533]: This was throwing an IndexError, "self.data_indices.internal_data.output.full" had indices larger than the current y shape

File: src/anemoi/training/data/dataset.py (hunk 5 above)
Comment (icedoom888) [reply src/anemoi/training/data/dataset.py:59]: In order to effectively reproduce other experiments (eg. Met-Norway) we need to follow the same training regime. 
So training with 10 epochs with bs=32 will require to know exactly the effective batch size to make the experiment reproducible.

File: src/anemoi/training/train/train.py (hunk 7 above)
Comment (icedoom888) [reply src/anemoi/training/train/train.py:45]: Yes sure! Here for debugging only

**Comment by icedoom888** [comment]
@mchantry with the current implementation the trainer is somehow skipping validation alltogether. 
I am looking into this, would you mind also taking a look to help me figure out where the issue is?

**Comment by gabrieloks** [comment]
@icedoom888  I can test this later today and see if validation works as intended

File: src/anemoi/training/train/forecaster.py (hunk 4 above)
Comment (mchantry) [reply src/anemoi/training/train/forecaster.py:533]: I am concerned that this is masking an issue with incompatible datasets.
Maybe this should be tested in conjunction with https://github.com/ecmwf/anemoi-training/pull/120

File: src/anemoi/training/train/train.py [hunk 8]
```diff
@@ -291,16 +315,13 @@ def _log_information(self) -> None:
...
-            / self.config.hardware.num_gpus_per_model
-        )
+        total_number_of_model_instances = self.config.hardware.num_nodes * self.config.hardware.num_gpus_per_node / self.config.hardware.num_gpus_per_model
+
         LOGGER.debug(
             "Total GPU count / model group size: %d - NB: the learning rate will be scaled by this factor!",
             total_number_of_model_instances,
         )
-        LOGGER.debug("Effective learning rate: %.3e", total_number_of_model_instances * self.config.training.lr.rate)
+        LOGGER.debug("Effective learning rate: %.3e", int(total_number_of_model_instances) * self.config.training.lr.rate)
```
Comment (mchantry) [inline src/anemoi/training/train/train.py:324]: Does it make sense to put the `int` here?

Comment (icedoom888) [reply src/anemoi/training/train/train.py:324]: I got an error at some point here.. i think it's a string?

**Comment by icedoom888** [comment]
All issues with validation, plotting and imputer are fixed. Cleaning it up before final review.

File: src/anemoi/training/train/forecaster.py (hunk 6 above)
Comment (icedoom888) [reply src/anemoi/training/train/forecaster.py]: Introduced all of this again ✔️

File: src/anemoi/training/train/forecaster.py (hunk 4 above)
Comment (icedoom888) [reply src/anemoi/training/train/forecaster.py]: Reverted this change ✔️

File: src/anemoi/training/train/train.py (hunk 7 above)
Comment (JesperDramsch) [reply src/anemoi/training/train/train.py:45]: In case you want to avoid setting these levels and forgetting, the documentation shows how to set these from the configs:

https://anemoi-training.readthedocs.io/en/latest/user-guide/debugging.html#debug-flags-for-better-error-handling

File: src/anemoi/training/data/dataset.py [hunk 9]
```diff
@@ -290,6 +293,10 @@ def __repr__(self) -> str:
             Timeincrement: {self.timeincrement}
         """
 
+    def __len__(self) -> int:
```
Comment (JesperDramsch) [inline src/anemoi/training/data/dataset.py:296]: We have encountered problems with setting a `__len__` attribute on the `IterableDataset` before. Is this thoroughly tested?

File: src/anemoi/training/train/train.py [hunk 10]
```diff
@@ -62,9 +64,14 @@ def __init__(self, config: DictConfig) -> None:
         OmegaConf.resolve(config)
         self.config = config
 
-        # Default to not warm-starting from a checkpoint
+        # Set Transfer Learning based on the other if not provided
+        if self.config.training.transfer_learning is None:
+            self.config.training.transfer_learning = (
```
Comment (JesperDramsch) [inline src/anemoi/training/train/train.py:69]: Why is the config value being overwritten here?

File: src/anemoi/training/utils/checkpoint.py [hunk 11]
```diff
@@ -65,3 +98,37 @@ def save_inference_checkpoint(model: torch.nn.Module, metadata: dict, save_path:
...
+    try:
+        checkpoint = torch.load(ckpt_path, map_location=model.device)
+
+    # TODO @anaprietonem: this is a patch for issue #57
+    except RuntimeError:
+        LOGGER.debug("Need to remove metadata from the checkpoint file due to issue #57..")
+
+        file_to_delete = "archive/anemoi-metadata/ai-models.json"
+        # Construct and execute the command
+        remove_file_from_zip(ckpt_path, file_to_delete)
```
Comment (JesperDramsch) [inline src/anemoi/training/utils/checkpoint.py:115]: The side effects from deleting provenance information are quite significant. While we need to find a solution to the problem, modifying a checkpoint in-place and removing information from it cannot go into `develop` like this.

**Review by JesperDramsch — DISMISSED** [review]

File: src/anemoi/training/utils/checkpoint.py [hunk 12]
```diff
@@ -65,3 +98,37 @@ def save_inference_checkpoint(model: torch.nn.Module, metadata: dict, save_path:
...
     return inference_filepath
+
+
+def transfer_learning_loading(model: torch.nn.Module, ckpt_path: Path | str) -> nn.Module:
+
+    # Load the checkpoint
+    try:
+        checkpoint = torch.load(ckpt_path, map_location=model.device)
+
+    # TODO @anaprietonem: this is a patch for issue #57
```
Comment (JesperDramsch) [inline src/anemoi/training/utils/checkpoint.py:109]: I'd probably avoid putting TODOs in other people's names.

File: src/anemoi/training/train/train.py (hunk 10 above)
Comment (icedoom888) [reply src/anemoi/training/train/train.py:69]: Tries to fill the value for old configs that have not been updated.
Loading a checkpoint using transfer learning on the same model and when not loading the optimizer state (self.config.training.load_weights_only=True) will actually not change the checpoint itself. Transfer learning becomes normal learning then.

The falg is actually not really necessary but it adds clarity.

File: src/anemoi/training/data/dataset.py (hunk 9 above)
Comment (icedoom888) [reply src/anemoi/training/data/dataset.py:296]: Have been using this for a while, did not encounter issues so far. You are welcome to test it out!

File: src/anemoi/training/utils/checkpoint.py (hunk 12 above)
Comment (icedoom888) [reply src/anemoi/training/utils/checkpoint.py:109]: This is just a note for Ana since she opened the issue. 
This way she will know to apply a change here when the issue is finally solved.
@anaprietonem Is this fine for you? :)

File: src/anemoi/training/utils/checkpoint.py (hunk 11 above)
Comment (icedoom888) [reply src/anemoi/training/utils/checkpoint.py:115]: My original implementation was creating a copy of the ckpt called "transfer.ckpt" where the provenance information is deleted. Should I re-implement this then?

Comment (icedoom888) [reply src/anemoi/training/utils/checkpoint.py]: Patching the checkpoint directly in the code caused file access issues when running in a multi GPU setting.
If someone has ideas we can reintorduce this in the future.
For the moment, **I had to remove this** and simplify the code.
Nowthe code is expecting that the checkpoint has been already fixed previously as documented in #57, and is present in the directory with the name: {checkpoint_path}/{checkpoint_name}_patched.ckpt

File: src/anemoi/training/utils/checkpoint.py [hunk 13]
```diff
@@ -7,16 +7,30 @@
...
 import torch
+import torch.nn as nn
 from anemoi.utils.checkpoints import save_metadata
 
 from anemoi.training.train.forecaster import GraphForecaster
 
+LOGGER = logging.getLogger(__name__)
+
+
+def create_new_zip_path(zip_path: Path | str, patch_str: str = "patched") -> Path:
```
Comment (mchantry) [inline src/anemoi/training/utils/checkpoint.py:24]: Sorry, I don't understand the role of this function. Can't a user simply provide the checkpoint name in the config file?

Comment (mchantry) [reply src/anemoi/training/utils/checkpoint.py:24]: Maybe I am missing something, sorry.

Comment (icedoom888) [reply src/anemoi/training/utils/checkpoint.py]: Hey Matt, talked to @gabrieloks and decided to ignore #57 on this and let the code crush for that issue.
This has been removed!

**Review by gabrieloks — APPROVED** [review]
Did some final tests and looks good to me!
